multiple fixes for implementation and tests on Windows with NTFS in German Germany US CEST

resident-uhlig opened the pull request:
Hello,

I wanted to help out and implement one feature request and follow the [build steps](https://h2database.com/html/build.html#building) as documented. However the setup of my development environment is rather unusual and so I stumbled over a lot of problems when running the tests. This pull requests should fix several issues. But there are still some tests breaking on my machine.

If you prefer to have each change in a separate PR please say so and I will split and re-order the commits.

Fixes and changes:
- added reminder that you have to re-compile using the build script before running the tests
- fixed URL to Maven repository: It must be SSL
- fixed `ToChar`: It was using the wrong names when Locale was changed at runtime. My environment starts at en-US for the `Category.FORMAT` but is changed to de-DE during the tests. I guess that is because my system's language is German, the location is Germany but the format is US.
- fixed two work-arounds for Windows where the Windows-API-error message is interpreted because it is localized and independent of Java's locale. To be honest this was a really dirty hack before my changes and still is. But I think there is no better way to do it. Please note this is only fixed for German (and English) systems. It will break for other languages.
- fixed reading last modified for `split:` files when lock mode is FILE (the file path contained `split:`)
- fixed tests deleting of read-only files on Windows by making them writable before
- fixed test for timezone's short name (H2's implementation always uses English for that)
- fixed Link Checker for windows (directory separtors for file system and for URLs are different)

Tests that still fail (list may be incomplete):
- TestOutOfMemory (always in every combination of TestAll)
- may be TestRecovery but probably fixed with split: lock files
- TestScript for "mvStore big"
- TestTransaction for "mvStore big"
- TestCrashAPI for "mvStore big"
- TestPowerOff for "mvStore big"
- TestReorderWrites for "mvStore big"
- may be performance test

I am looking forward to your comments and sorry for the messy commit history.

I wrote the code, it's mine, and I'm contributing it to H2 for distribution multiple-licensed under the MPL 2.0, and the EPL 1.0 (https://h2database.com/html/license.html). Of course I copy & pasted the windows error messages.

resident-uhlig committed:
fix: timezone and locale issues in tests

resident-uhlig committed:
docs: `build compile` is required before `build test`

resident-uhlig committed:
fix: link to maven.org requires HTTPS

resident-uhlig committed:
fix: some filesystems cannot delete read-only files

resident-uhlig committed:
fix: reload names in ToChar when Locale for FORMAT changes

resident-uhlig committed:
fix: localized IOException on windows

the full German text is
java.io.IOException: Der Vorgang ist bei einer Datei mit einem geöffneten Bereich, der einem Benutzer zugeordnet ist, nicht anwendbar

resident-uhlig committed:
style: line in comment too long

resident-uhlig committed:
fix: windows uses backslash, URLs use slash

resident-uhlig committed:
chore: make SpellChecker happy

resident-uhlig committed:
fix: more windows messages

resident-uhlig committed:
fix: aggressiveLastModified for split: files

resident-uhlig commented:
The build has failed for [openjdk11](https://travis-ci.org/github/h2database/h2database/jobs/689847139)
```plain
23:32:32 00:00.615 org.h2.test.db.TestFunctions FAIL java.lang.AssertionError: Expected: 8:12:34 A<*>M (10) actual: 8:12:34 vorm. (13) 
ERROR: FAIL (mvStore memory ) java.lang.AssertionError: Expected: 8:12:34 A<*>M (10) actual: 8:12:34 vorm. (13)  java.lang.AssertionError: Expected: 8:12:34 A<*>M (10) actual: 8:12:34 vorm. (13) 
```

"vorm." is the closest thing to "AM" (ante meridiem) in German. I will look into that.

Interestingly it passed for [oraclejdk8](https://travis-ci.org/github/h2database/h2database/jobs/689847140)

katzyn commented:
1. `https:` for Maven is OK.
2. Old code that uses text of messages is a hack and we're not going to make this hack even bigger. Please, remove `WindowsLocalizedMessages` from this pull request and create a new issue for your problem for discussion. Maybe somebody find a better idea how to fix this problem for all locales.
3. I think `setWritable()` should be used before deletion on Windows only, add an OS check for it. On other systems it only increases I/O without any good reason.
4. `FilePath.setWritable()` doesn't properly work on POSIX systems in all cases and its limitations at least should be documented. User of H2's process and owner of the file can be different, for example. Hypothetically you can simply move the code for Windows to `delete()` methods to reduce amount of unreliable code.
5. Code in `ToChar` is a hack. I know about the problem on some locales, but it doesn't look like the proper way to fix it. A locale of the session should be used instead, but it isn't yet implemented. Currently we use the initial system locale of Java process and I think we should continue to use it until session's locale will be introduced.

resident-uhlig committed:
Revert "fix: more windows messages"

This reverts commit 87b1dcedb30617f6e114d3d28fbab57c7744f8e2.

resident-uhlig committed:
Revert "chore: make SpellChecker happy"

This reverts commit b09b0d0102e483356082c49b44b02102605f4052.

resident-uhlig committed:
Revert "fix: localized IOException on windows"

This reverts commit ab5cded12b98f0a7d3384575c4350598df659e98.

resident-uhlig committed:
refactor: set writable before delete only on Windows

resident-uhlig committed:
Revert "fix: reload names in ToChar when Locale for FORMAT changes"

This reverts commit d2f5a0ba2d668251d273acc46fc8215305b23fad.

resident-uhlig commented:
> Please, remove `WindowsLocalizedMessages` from this pull request and create a new issue for your problem for discussion.

I have reverted the commits and created issue #2638 

> 3. I think `setWritable()` should be used before deletion on Windows only, add an OS check for it. [...]
> 4. `FilePath.setWritable()` doesn't properly work on POSIX systems in all cases and its limitations at least should be documented.

I have refactored the code to be only called if it is run on Windows and removed `FilePath.setWritable()`.

> 5. Code in `ToChar` is a hack. [...] it doesn't look like the proper way to fix it. A locale of the session should be used instead, but it isn't yet implemented.

I have reverted the commits. Also I have set my OS to English language and the respective tests work now. Should I create a new issue for this? There is the issue #2072 but it doesn't quite fit because it is about the language of the H2 console, not H2's formatting of database values.

katzyn commented:
#2072 is expected to be resolved with introduction of session's locale, but it's a large piece of work because many things need to be reimplemented. After such change `ToChar` and some other functions can be switched to use locale of the session.
